### PR TITLE
[5.3] Add partition collection method

### PR DESCRIPTION
**Pull request by freekmurze**
The partition method returns a collection with two elements. Items in the first element did pass the given `$callback`, items in the second element did not.

Here's an example:

```php
collect(range(1, 10))->partition(function ($i) {
    return $i <= 5;
});

// returns a new collection
// first element is a collection with items 1, 2 ,3, 4, 5
// second element is a collection with items 5, 6, 7, 8, 9, 10
```

### CODE REVIEW

File: src/Illuminate/Support/Collection.php [hunk 1]
```diff
@@ -726,6 +726,24 @@ public function forPage($page, $perPage)
...
+     */
+    public function partition(callable $callback)
+    {
+        $partitions = [new static(), new static()];
+
+        foreach ($this->items as $item) {
+            $partitions[! (int) $callback($item)][] = $item;
+        }
+
+        return new static($partitions);
```
Comment (taylorotwell) [inline src/Illuminate/Support/Collection.php]: Shouldn't you just return the plain array so you can separate them using `lists`?

**Comment by GrahamCampbell** [comment]
Looks cool. It should be noted that the strict definition of a partition requires it to be non-empty, though, I guess it's probably ok in practical terms.

File: src/Illuminate/Support/Collection.php [hunk 2]
```diff
@@ -726,6 +726,24 @@ public function forPage($page, $perPage)
     }
 
     /**
+     * Returns an array with two elements. Items in the first element did pass
```
Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: Please format like the other phpdoc, then looks good. 👍

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Support/Collection.php (hunk 2 above)
Comment (freekmurze) [inline src/Illuminate/Support/Collection.php]: I'm afraid I don't see the formatting problem. Could you tell what's wrong with the current formatting?

Comment (GrahamCampbell) [inline src/Illuminate/Support/Collection.php]: Short descriptions must be "sort". Exactly one sentence and one line. The rest must be seperated.

Comment (freekmurze) [inline src/Illuminate/Support/Collection.php]: Got it, thanks for clarifying

File: src/Illuminate/Support/Collection.php (hunk 1 above)
Comment (JosephSilber) [inline src/Illuminate/Support/Collection.php]: Actually, `list` works on collections, so I think we should definitely return a collection.

Really no collection method should ever return a plain array. You don't ever want to break the chain.

**Pull request merged**